core/src/test/java/org/infinispan/distexec/DistributedExecutorFailoverTest.java

```diff
@@ -0,0 +1,126 @@
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.manager.EmbeddedCacheManager;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.test.MultipleCacheManagersTest;
+import org.infinispan.test.TestingUtil;
+import org.testng.AssertJUnit;
+import org.testng.annotations.Test;
+
+/**
+ * Tests dist.exec failover after node kill.
+ *
+ * @author Vladimir Blagojevic
+ * @since 7.1
+ */
+@Test(groups = "functional", testName = "distexec.DistributedExecutorFailoverTest")
+public class DistributedExecutorFailoverTest extends MultipleCacheManagersTest {
+
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      ConfigurationBuilder builder = getDefaultClusteredCacheConfig(CacheMode.DIST_SYNC, false);
+      addClusterEnabledCacheManager(builder);
+      addClusterEnabledCacheManager(builder);
+      addClusterEnabledCacheManager(builder);
+
+      waitForClusterToForm(cacheName());
+   }
+
+   protected String cacheName() {
+      return "DistributedExecutorFailoverTest";
+   }
+
+   public void testBasicTargetRemoteDistributedCallable() throws Exception {
+      long taskTimeout = TimeUnit.SECONDS.toMillis(15);
+      long sleepTaskTimeout = taskTimeout / 4;
+      EmbeddedCacheManager cacheManager1 = manager(0);
+      final EmbeddedCacheManager cacheManager2 = manager(1);
+
+      Cache<Object, Object> cache1 = cacheManager1.getCache();
+      Cache<Object, Object> cache2 = cacheManager2.getCache();
+      DistributedExecutorService des = null;
+      try {
+         // initiate task from cache1 and execute on same node
+         des = new DefaultExecutorService(cache1);
+         Address target = cache2.getAdvancedCache().getRpcManager().getAddress();
+
+         DistributedTaskBuilder<Integer> builder = des
+               .createDistributedTaskBuilder(new SleepingSimpleCallable(sleepTaskTimeout))
+               .failoverPolicy(new RandomNodeTaskFailoverPolicy(2))
+               .timeout(taskTimeout, TimeUnit.MILLISECONDS);
+
+         Future<Integer> future = des.submit(target, builder.build());
+         fork(new Runnable() {
+
+            @Override
+            public void run() {
+               TestingUtil.killCacheManagers(cacheManager2);
+            }
+         });
+
+         AssertJUnit.assertEquals((Integer) 1, future.get());
+      } catch (Exception ex) {
+         AssertJUnit.fail("Task did not failover properly");
+      } finally {
+         des.shutdown();
+      }
+   }
+
+   static class SleepingSimpleCallable implements Callable<Integer>, Serializable {
+
+      /**
+       *
+       */
+      private static final long serialVersionUID = 6189005766285399200L;
+      private long timeout;
+
+      public SleepingSimpleCallable(long timeout) {
+         super();
+         this.timeout = timeout;
+      }
+
+      @Override
+      public Integer call() throws Exception {
+         Thread.sleep(timeout);
```

Comment:
Right, but sleeps like this are sometime the cause why tests fail randomly. Plus, what is the exact assumption of this test in terms of events? I guess `killCacheManagers` should be called while the callable is sleeping? What would happen if GC hit the testsuite and the callable finished (sleep finished) before `killCacheManagers` was called?

Dan's suggestion helps answers these questions and should make the test more predictable.